mm2src/mm2_bitcoin/keys/src/address.rs

```diff
@@ -26,19 +28,61 @@ pub enum Type {
     P2SH,
 }
 
+#[derive(Clone, Debug, Display, Deserialize, PartialEq, Serialize)]
+#[serde(tag = "format")]
+pub enum AddressFormat {
+    /// Standard UTXO address format.
+    /// In Bitcoin Cash context the standard format also known as 'legacy'.
+    #[serde(rename = "standard")]
+    #[display(fmt = "Legacy")]
+    Standard,
+    /// Segwit Address
+    /// https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki
+    #[serde(rename = "segwit")]
+    Segwit,
+    /// Bitcoin Cash specific address format.
+    /// https://github.com/bitcoincashorg/bitcoincash.org/blob/master/spec/cashaddr.md
+    #[serde(rename = "cashaddress")]
+    #[display(fmt = "CashAddress")]
+    CashAddress {
+        network: String,
+        #[serde(default)]
+        pub_addr_prefix: u8,
+        #[serde(default)]
+        p2sh_addr_prefix: u8,
+    },
+}
+
+impl Default for AddressFormat {
+    fn default() -> Self { AddressFormat::Standard }
+}
+
+impl AddressFormat {
+    pub fn is_segwit(&self) -> bool { matches!(*self, AddressFormat::Segwit) }
+
+    pub fn is_cashaddress(&self) -> bool { matches!(*self, AddressFormat::CashAddress { .. }) }
+
+    pub fn is_legacy(&self) -> bool { matches!(*self, AddressFormat::Standard) }
+}
+
 /// `AddressHash` with prefix and t addr zcash prefix
 #[derive(Debug, PartialEq, Clone)]
 pub struct Address {
     /// The prefix of the address.
     pub prefix: u8,
     /// T addr prefix, additional prefix used by Zcash and some forks
     pub t_addr_prefix: u8,
+    /// Segwit addr human readable part
+    pub hrp: Option<String>,
     /// Public key hash.
     pub hash: AddressHash,
     /// Checksum type
     pub checksum_type: ChecksumType,
+    /// Address Format
+    pub addr_format: AddressFormat,
 }
 
+// Todo: add segwit checksum detection
```

Comment:
You could miss this TODO :)

Reply:
I discussed this with @artemii235, I think this will be a separate issue to refactor the address code just added the TODO to not forget :) 
also, the checksum for Segwit addresses gets checked here for now https://github.com/KomodoPlatform/atomicDEX-API/blob/e97e2fc32315ba54d8bc4adc0ce24f7820d1dbcc/mm2src/mm2_bitcoin/keys/src/segwitaddress.rs#L153

Reply:
Got it, thanks!